service/src/main/java/crawlercommons/urlfrontier/service/AbstractFrontierService.java

```diff
@@ -941,14 +946,23 @@ public void listURLs(
                 CloseableIterator<URLItem> urliter = urlIterator(e);
 
                 while (urliter.hasNext()) {
-                    totalCount++;
-                    if (totalCount < start) {
-                        urliter.next();
-                    } else if (sentCount < maxURLs) {
-                        responseObserver.onNext(urliter.next());
-                        sentCount++;
-                    } else {
-                        break;
+                    URLItem cur = urliter.next();
+
+                    if (StringUtils.isEmpty(filter)
+                            || (!ignoreCase && cur.getKnown().getInfo().getUrl().contains(filter))
+                            || (ignoreCase
+                                    && StringUtils.containsIgnoreCase(
+                                            cur.getKnown().getInfo().getUrl(), filter))) {
+
+                        if (totalCount < start) {
```

Comment:
IIRC that's why totalCount was set to -1 and not 0; if you change to 0 then probably totalCount <= start